### PR TITLE
Fix race between jvm shutdown and `writer.finished`

### PR DESCRIPTION
**Pull request by ttim**
Currently `writer.finished` happens in `onComplete` callback on `Future` result in `Exucution`. However since `onComplete` isn't being called before future being resolved and called asynchroniously after future being resolved, it leads to a race and runtime error:
- User's code as last operation in `main` executes `Execution`
- `onComplete` with `writer.finished` is being scheduled
- result `Future` gets resolved and jvm starts to shutdown
- `writer.finished` starts to execute and in case of cascading backend adds shutdown hook
- which is not permitted during jvm shutdown and breaks

To fix this behaviour I made `onComplete` logic to happen before result future get resolved by changing `onComplete` to `andThen`

### CODE REVIEW

**Comment by tlazaro** [comment]
LGTM, both onComplete vs andThen are for performing side-effects when the Future is done. But andThen makes it so the side-effect runs before anything else. Seems like exactly what we wanted in this case.

**Comment by ttim** [comment]
@johnynek does it seem reasonable to you? I'm wondering is there something I missing

File: scalding-core/src/main/scala/com/twitter/scalding/Execution.scala [hunk 1]
```diff
@@ -155,9 +155,8 @@ sealed trait Execution[+T] extends Serializable { self: Product =>
     val exec = Execution.optimize(conf, this)
     // get on Trampoline
     val CFuture(fut, cancelHandler) = exec.runStats(confWithId, mode, ec)(cec).get
-    val result = fut.map(_._1)
     // When the final future in complete we stop the submit thread
-    result.onComplete { t =>
+    val result = fut.map(_._1).andThen { case t =>
       if (t.isFailure) {
         // cancel running executions if this was a failure
         Await.ready(cancelHandler.stop(), scala.concurrent.duration.Duration(30, scala.concurrent.duration.SECONDS))
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/Execution.scala]: I think we are technically supposed to do:
```scala
blocking {
  Await.ready(cancelHandler.stop(), ...)
```
here:

https://docs.scala-lang.org/overviews/core/futures.html#blocking-inside-a-future

Other than that, it looks good.

Small nit, we should probably import Duration and SECOND from `scala.concurrent.duration`

Comment (ttim) [reply scalding-core/src/main/scala/com/twitter/scalding/Execution.scala]: Can't import `Duration` because of scalding `Duration`, imported `scala.concurrent.duration` instead.

Comment (ttim) [reply scalding-core/src/main/scala/com/twitter/scalding/Execution.scala]: Added blocking block around `Await`

File: scalding-parquet/src/test/java/com/twitter/scalding/parquet/example/TestInputOutputFormat.java [hunk 2]
```diff
@@ -79,7 +79,7 @@
   private static final Logger LOG = LoggerFactory.getLogger(TestInputOutputFormat.class);
 
   final Path parquetPath = new Path("target/test/example/TestInputOutputFormat/parquet");
-  final Path inputPath = new Path("src/test/java/org/apache/parquet/hadoop/example/TestInputOutputFormat.java");
```
Comment (johnynek) [inline scalding-parquet/src/test/java/com/twitter/scalding/parquet/example/TestInputOutputFormat.java:82]: Why this change?

**Review by johnynek — APPROVED** [review]
One minor question but other than that this looks good.

File: scalding-parquet/src/test/java/com/twitter/scalding/parquet/example/TestInputOutputFormat.java (hunk 2 above)
Comment (ttim) [reply scalding-parquet/src/test/java/com/twitter/scalding/parquet/example/TestInputOutputFormat.java:82]: This test was copied by @tlazaro in previous PR, and it relies on reading itself to run properly. Since path of copied file is different it started to fail, so I changed it.

**Review by johnynek — APPROVED** [review]
